File: src/main/java/com/lylx/learn/singletonInJava/LazyThreadSafeSynchronized.java
package com.lylx.learn.singletonInJava;

/**
 * 懒汉式，线程安全：同步锁（开销大）
 */
public class LazyThreadSafeSynchronized {
    private static LazyThreadSafeSynchronized INSTANCE;

    private LazyThreadSafeSynchronized(){}

    public static synchronized LazyThreadSafeSynchronized getInstance(){
        if(INSTANCE == null){
            INSTANCE = new LazyThreadSafeSynchronized();
        }
        return INSTANCE;
    }
}

File: src/main/java/com/lylx/learn/singletonInKotlin/LazyThreadSafeSynchronized.kt
package com.lylx.learn.singletonInKotlin

class LazyThreadSafeSynchronized private constructor(){
    companion object {
        private var instance: LazyThreadSafeSynchronized? = null

        @Synchronized
        fun get(): LazyThreadSafeSynchronized{
            if(instance == null) instance = LazyThreadSafeSynchronized()
            return instance!!
        }
    }
}